File: gulo-fe/src/api/cache/s3-cache.ts
import { GetObjectCommand, PutObjectCommand, S3Client } from '@aws-sdk/client-s3';
import dotenv from 'dotenv';
import { toast } from 'sonner';

dotenv.config();

const s3Client = new S3Client({
  region: process.env.S3_REGION,
  credentials: {
    accessKeyId: process.env.ACCESS_KEY_ID!,
    secretAccessKey: process.env.SECRET_ACCESS_KEY!,
  },
});

const BUCKET_NAME = process.env.S3_BUCKET_NAME;

const streamToString = async (stream: ReadableStream<Uint8Array>): Promise<string> => {
  const reader = stream.getReader();
  const utf8Decoder = new TextDecoder('utf-8');
  let result = '';
  let exitLoop = false;

  while (!exitLoop) {
    const { done, value } = await reader.read();
    if (done) {
      exitLoop = true;
      break;
    }
    result += utf8Decoder.decode(value, { stream: true });
  }

  result += utf8Decoder.decode();
  return result;
};

export const getCacheFromS3 = async (key: string): Promise<Record<string, any> | null> => {
  try {
    const params = { Bucket: BUCKET_NAME!, Key: key };
    const command = new GetObjectCommand(params);
    const data = await s3Client.send(command);

    const bodyContents = await streamToString(data.Body as ReadableStream<Uint8Array>);
    return JSON.parse(bodyContents);
  } catch (error: unknown) {
    if (error instanceof Error && error.name === 'NoSuchKey') {
      toast.error("Oopsie, let's try refreshing the page, shall we?");
      await saveCacheToS3(key, {});
      return {};
    }
    toast.error("Oopsie, let's try refreshing the page, shall we?");
    return null;
  }
};

export const saveCacheToS3 = async (key: string, cache: Record<string, any>): Promise<void> => {
  try {
    const params = {
      Bucket: BUCKET_NAME!,
      Key: key,
      Body: JSON.stringify(cache),
      ContentType: 'application/json',
    };
    const command = new PutObjectCommand(params);
    await s3Client.send(command);
  } catch (error: unknown) {
    toast.error("Oopsie, let's try refreshing the page, shall we?");
  }
};

File: gulo-be/graph_compute_lambda/s3_cache.py
import os
import json
import boto3
import logging

logger = logging.getLogger()
logger.setLevel(logging.INFO)

aws_access_key_id = os.getenv("ACCESS_KEY_ID")
aws_secret_access_key = os.getenv("SECRET_ACCESS_KEY")
s3_region = os.getenv("S3_REGION", "eu-central-1")

s3_client = boto3.client(
    "s3",
    aws_access_key_id=aws_access_key_id,
    aws_secret_access_key=aws_secret_access_key,
    region_name=s3_region,
)


def get_cache_from_s3(key):
    try:
        response = s3_client.get_object(Bucket=os.environ["S3_BUCKET_NAME"], Key=key)
        return json.loads(response["Body"].read().decode("utf-8"))
    except s3_client.exceptions.NoSuchKey:
        logger.info("No cache found for key: %s", key)
        return {
            "count": 0,
            "streams": [],
        }
    except Exception as e:
        logger.error("Error fetching cache from S3: %s", e)
        raise


def save_cache_to_s3(key, data):
    try:
        s3_client.put_object(
            Bucket=os.environ["S3_BUCKET_NAME"], Key=key, Body=json.dumps(data)
        )
        logger.info("Cache saved to S3 at key: %s", key)
    except Exception as e:
        logger.error("Error saving cache to S3: %s", e)
        raise
